pootle/apps/pootle_fs/finder.py

```diff
@@ -0,0 +1,182 @@
+        if "<dir_path>" in path:
+            if dir_path and dir_path.strip("/"):
+                path = path.replace(
+                    "<dir_path>", "/%s/" % dir_path.strip("/"))
+            else:
+                path = path.replace("<dir_path>", "")
+        local_path = path.replace(self.file_root, "")
+        if "//" in local_path:
+            path = os.path.join(
+                self.file_root,
+                local_path.replace("//", "/").lstrip("/"))
+        return path
+
+    def _ext_re(self):
+        return (
+            r".(?P<ext>(%s))"
+            % "|".join(
+                ("%s$" % x)
+                for x in set(self.ext)))
+
+    def _parse_path_regex(self):
+        path = self.translation_path
+        for k, v in self.path_mapping:
+            path = path.replace(k, v)
+        return "%s%s$" % (
+            os.path.splitext(path)[0],
+            self._ext_re())
+
+    def _parse_path(self):
+        regex = self._parse_path_regex()
+        if not self.path_filters:
+            return regex
+        regex = r"(?=%s)" % regex
+        filter_regex = "".join(
+            [("(?=^%s)" % fnmatch.translate(path))
+             for path in self.path_filters])
+        return "%s%s" % (regex, filter_regex)
+
+
+class TranslationPathValidator(object):
+
+    validators = (
+        "no_dots", "lang_code", "ext", "match_tags", "path")
+
+    def __init__(self, path):
+        self.path = path
+
+    def validate_no_dots(self):
+        links_to_parent = (
+            self.path == ".."
+            or self.path.startswith("../")
+            or "/../" in self.path
+            or self.path.endswith("/.."))
+        if links_to_parent:
+            raise ValueError(
+                "Translation path should not contain '..'")
+
+    def validate_lang_code(self):
+        if "<language_code>" not in self.path:
+            raise ValueError(
+                "Translation path must contain a <language_code> pattern to match.")
+
+    def validate_ext(self):
+        if not self.path.endswith(".<ext>"):
+            raise ValueError(
+                "Translation path must end with ext.")
+
+    @cached_property
+    def stripped_path(self):
+        return (self.path.replace("<language_code>", "")
+                         .replace("<dir_path>", "")
+                         .replace("<ext>", "")
+                         .replace("<filename>", ""))
+
+    def validate_match_tags(self):
+        if "<" in self.stripped_path or ">" in self.stripped_path:
+            raise ValueError(
+                "Only <language_code>, <dir_path>, <filename> and <ext> are valid "
+                "patterns to match in the translation path")
+
+    def validate_path(self):
+        if re.search("[^\w\/\-\.]+", self.stripped_path):
+            raise ValueError(
+                "Translation paths can only contain alpha-numeric characters, "
+                "_ or -: '%s'" % self.path)
```

Comment:
If you can find out which are the offending characters it looks like a better approach.